Data Structure 2 Assignment/Counseling Process/src/com/metacube/facade/CounselingFacade.java

```diff
@@ -0,0 +1,168 @@
+ */
+public class CounselingFacade {
+	private MyQueue<Candidate> candidateQueue;
+	private static CounselingFacade counselingFacade;
+	private List<Candidate> candidateList;
+
+	/**
+	 * Instantiates a new counseling facade.
+	 */
+	private CounselingFacade() {
+		initializeCandidateQueue();
+	}
+
+	/**
+	 * Gets the single instance of CounselingFacade.
+	 *
+	 * @return single instance of CounselingFacade
+	 */
+	public static CounselingFacade getInstance() {
+		if (counselingFacade == null) {
+			synchronized (CounselingFacade.class) {
+				if (counselingFacade == null) {
+					counselingFacade = new CounselingFacade();
+				}
+			}
+		}
+
+		return counselingFacade;
+	}
+
+	/**
+	 * Initialize candidate queue.
+	 */
+	private void initializeCandidateQueue() {
+		candidateQueue = new MyQueue<Candidate>();
+		candidateList = CandidateInMemoryDao.getInstance().getCandidatesList();
+		Collections.sort(candidateList);
+		for (Candidate candidate : candidateList) {
+			candidateQueue.enqueue(candidate);
+		}
+	}
+
+	/**
+	 * Gets the college capacity map.
+	 *
+	 * @return the college capacity map between college id and capacity
+	 */
+	private Map<String, Integer> getCollegeCapacityMap() {
+		Map<String, Integer> capacityMap = new HashMap<String, Integer>();
+		List<College> collegeList = CollegeInMemoryDao.getInstance()
+				.getCollegesList();
+		for (College college : collegeList) {
+			capacityMap.put(college.getId(), college.getSeatsAvailable());
+		}
+
+		return capacityMap;
+	}
+
+	/**
+	 * Start counseling.
+	 */
+	public void startCounseling() {
+		CollegeInMemoryDao collegeDao = CollegeInMemoryDao.getInstance();
+		Map<String, Integer> collegeMap = getCollegeCapacityMap();
+		String choice;
+		boolean skip;
+		boolean seatAlloted;
+		int count = candidateQueue.size();
+		// condition for total candidates each counseling round run for
+		// exact one time for each candidate
+		while (count != 0) {
+			seatAlloted = false;
+			skip = false;
+			Candidate candidate = (Candidate) candidateQueue.dequeue();
+			String[] candidateChoice = candidate.getOptedColleges();
+			// go through each choice
+			for (int index = 0; index < candidateChoice.length; index++) {
+				// check for the college is seat available or not
+				// if available then
+				if (collegeMap.get(candidateChoice[index]) != 0) {
+					choice = Controller.getInstance().getInput(
+							candidate.getName(),
+							collegeDao.getCollegeById(candidateChoice[index])
+									.getName());
+					if (choice.equalsIgnoreCase("y")) {
+						candidate.setCollegeId(candidateChoice[index]);
+						collegeDao
+								.getCollegeById(candidateChoice[index])
+								.setSeatsAvailable(
+										collegeMap.get(candidateChoice[index]) - 1);
+						seatAlloted = true;
+						skip = false;
+					} else {
+						skip = true;
+						// if candidate reject first choice then he not eligible
+						// for next round
+						if (index == 0) {
```

Comment:
if candidate reject first choice then he not eligible, but still it should be available in list, so that it can be printed as non college assigned student.